### PR TITLE
Machinery to fix total issuance accounting and keep it fixed

### PR DESCRIPTION
**Pull request by orriin**
# DO NOT MERGE UNTIL CONST HAS APPROVED

Machinery for #563 

- Adds documentation for `pallet_subtensor::TotalIssuance`
- Introduces new storage item `TotalSubnetLocked` to track the total amount of tokens locked for subnet registration
- Introduces migration to initialize `TotalSubnetLocked` and then rejig the total issuance to be the sum of issued tokens, staked tokens, and tokens locked for subnet registration
- Adjusts `set_subnet_locked_balance` to update `TotalSubnetLocked` to reflect changes in `SubnetLocked`
- Creates custom `on_nonzero_unbalanced` handler for transaction fees, which will remove the burnt transaction fee amount from `pallet_subtensor::TotalIssuance` before finally dropping it
  - Tested on localnet 
- Adds `try-state` checks for
  - `TotalStake`
  - `TotalSubnetLock`
  - `TotalIssuance`

### CODE REVIEW

File: pallets/admin-utils/src/lib.rs
Comment (open-junius) [inline pallets/admin-utils/src/lib.rs:29]: For TotalSubnetLocked, we can set it via a function as other hyperparameter. Why introduce the pallet_subtensor::Config this time. is anything special?

Comment (open-junius) [reply pallets/admin-utils/src/lib.rs:29]: If it is a temp solution, we can easily remove it later?

File: pallets/subtensor/src/lib.rs
Comment (open-junius) [inline pallets/subtensor/src/lib.rs]: Can we put it in on_finalize, then it can update total issuance automatically

File: pallets/admin-utils/src/lib.rs
Comment (orriin) [reply pallets/admin-utils/src/lib.rs:29]: It's the idiomatic way to achieve this type of pallet coupling in Substrate, it allows us to directly access items from the subtensor pallet without needing all the boilerplate of specifying and implementing the interface. Notice I didn't need any additional boilerplate in my changes here, I just directly access pallet_subtensor storage items. I'll be doing a show and tell on this in a couple weeks :)

File: pallets/subtensor/src/lib.rs
Comment (orriin) [reply pallets/subtensor/src/lib.rs]: No need, we only need to call it once. After that, it will stay synced on its own (unless there's an edge case I missed 😅).

File: pallets/subtensor/src/lib.rs
Comment (camfairchild) [inline pallets/subtensor/src/lib.rs]:
```suggestion
    /// It is comprised of three parts:
```

File: pallets/subtensor/src/lib.rs
Comment (camfairchild) [inline pallets/subtensor/src/lib.rs]:
```suggestion
    /// The total amount of locked tokens in subnets for subnet registration.
```

**Review by open-junius — DISMISSED** [review]

**Review by distributedstatemachine — DISMISSED** [review]

File: pallets/subtensor/src/lib.rs
Comment (camfairchild) [inline pallets/subtensor/src/lib.rs]:
```suggestion
    /// separate accounting.
```

File: pallets/subtensor/src/utils.rs
Comment (camfairchild) [inline pallets/subtensor/src/utils.rs]:
```suggestion
        // Deduct the previous amount and add the new amount to the total
```

**Review by camfairchild — COMMENTED** [review]
Do we have a test for the total issuance fees hook? If not, that would be great to add.

**Comment by orriin** [comment]
> Do we have a test for the total issuance fees hook? If not, that would be great to add.

It will be tested every pr with live state in the existing try-runtime tests

**Review by open-junius — DISMISSED** [review]

**Review by camfairchild — DISMISSED** [review]

**Review by sam0x17 — DISMISSED** [review]

File: pallets/subtensor/src/lib.rs
Comment (distributedstatemachine) [inline pallets/subtensor/src/lib.rs:2338]: I like this alot . We should probably add more maps to this.

File: runtime/src/migrations/initialise_ti.rs
Comment (distributedstatemachine) [inline runtime/src/migrations/initialise_ti.rs]: What is the advantage of having this migration in the runtime? I find it confusing that we have some migrations in the pallet , and some in the runtime. Can we move it to the subtensor module?

**Review by distributedstatemachine — CHANGES_REQUESTED** [review]

File: pallets/admin-utils/src/lib.rs
Comment (distributedstatemachine) [inline pallets/admin-utils/src/lib.rs]: Remove

File: pallets/admin-utils/tests/tests.rs
Comment (distributedstatemachine) [inline pallets/admin-utils/tests/tests.rs]: Remove

File: pallets/admin-utils/tests/tests.rs
Comment (distributedstatemachine) [inline pallets/admin-utils/tests/tests.rs]: Remove all tests

**Review by distributedstatemachine — CHANGES_REQUESTED** [review]
Can we also add tests for the migrations ?

File: runtime/src/migrations/initialise_ti.rs
Comment (orriin) [reply runtime/src/migrations/initialise_ti.rs:7]: They should be in the runtime actually, it allows pre/post checks and better integration with try-runtime-cli.

**Comment by orriin** [comment]
DO NOT MERGE UNTIL CONST HAS APPROVED

File: runtime/src/migrations/initialise_ti.rs
Comment (orriin) [reply runtime/src/migrations/initialise_ti.rs]: I moved the migration definition to the pallet.

It's best practice for the migrations to be passed via the runtime so that try-runtime-cli can run pre/post hooks, do addition checks, and give better error messages.

**Pull request closed without merging**
